compile error with vulkan

##### System information (version)
- OpenCV => 4.0
- Operating System / Platform => Ubuntu 16.04
- Compiler => clang++

##### Detailed description

I try to compile opencv 4.0 with vulkan for android. When I compile them, there is an error related to vulkan.

../modules/dnn/src/vkcom/src/op_base.cpp:48:43: error: variable-sized object may not be initialized
    VkDescriptorSetLayoutBinding bindings[buffer_num] = {};
                                          ^~~~~~~~~~

build script is 
[28/144] Building CXX object modules/d...cv_dnn.dir/src/vkcom/src/op_base.cpp.o
FAILED: /home/chan/Android/Sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=armv7-none-linux-androideabi16 --gcc-toolchain=/home/chan/Android/Sdk/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64 --sysroot=/home/chan/Android/Sdk/ndk-bundle/sysroot  -DCVAPI_EXPORTS -DDISABLE_POSIX_MEMALIGN -DHAVE_PROTOBUF=1 -DTH_DISABLE_HEAP_TRACKING -DVK_NO_PROTOTYPES -D_USE_MATH_DEFINES -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I. -I../3rdparty/include -I../modules/dnn/include -Imodules/dnn -I../modules/core/include -I../modules/imgproc/include -I../modules/dnn/misc/caffe -I../modules/dnn/misc/tensorflow -I../modules/dnn/misc/onnx -isystem ../3rdparty/protobuf/src -isystem /home/chan/Android/Sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include -isystem /home/chan/Android/Sdk/ndk-bundle/sources/android/support/include -isystem /home/chan/Android/Sdk/ndk-bundle/sources/cxx-stl/llvm-libc++abi/include -isystem /home/chan/Android/Sdk/ndk-bundle/sysroot/usr/include/arm-linux-androideabi -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -mfpu=neon -Wa,--noexecstack -Wformat -Werror=format-security -std=c++11    -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security   -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith  -Wsign-promo -Wuninitialized -Winit-self -Winconsistent-missing-override -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -fdiagnostics-show-option -Qunused-arguments  -fvisibility=hidden -fvisibility-inlines-hidden -Wno-deprecated -Wno-missing-prototypes -Wno-missing-declarations -Wno-shadow -Wno-unused-parameter -Wno-sign-compare -Wno-invalid-offsetof -Oz -DNDEBUG   -DNDEBUG -fPIC   -std=c++11 -MD -MT modules/dnn/CMakeFiles/opencv_dnn.dir/src/vkcom/src/op_base.cpp.o -MF modules/dnn/CMakeFiles/opencv_dnn.dir/src/vkcom/src/op_base.cpp.o.d -o modules/dnn/CMakeFiles/opencv_dnn.dir/src/vkcom/src/op_base.cpp.o -c ../modules/dnn/src/vkcom/src/op_base.cpp

--- a/modules/dnn/src/vkcom/include/op_pool.hpp
+++ b/modules/dnn/src/vkcom/include/op_pool.hpp
@@ -59,7 +59,7 @@ private:
     int avg_pool_padded_area_;
     int need_mask_;
     PaddingMode padding_mode_;
-    int activation_;
+    //int activation_;
     PoolShaderConfig config_;
 };
 

--- a/modules/dnn/src/vkcom/src/context.cpp
+++ b/modules/dnn/src/vkcom/src/context.cpp
@@ -281,10 +281,14 @@ Context::~Context()
     if (enableValidationLayers) {
         auto func = (PFN_vkDestroyDebugReportCallbackEXT)
             vkGetInstanceProcAddr(kInstance, "vkDestroyDebugReportCallbackEXT");
-        if (func == nullptr) {
-            throw std::runtime_error("Could not load vkDestroyDebugReportCallbackEXT");
+        if (func == nullptr)
+        {
+            CV_LOG_FATAL(NULL, "Could not load vkDestroyDebugReportCallbackEXT");
         }
-        func(kInstance, kDebugReportCallback, NULL);
+        else
+        {
+            func(kInstance, kDebugReportCallback, NULL);
+        }
     }
     kShaders.clear();
     vkDestroyInstance(kInstance, NULL);

--- a/modules/dnn/src/vkcom/src/context.hpp
+++ b/modules/dnn/src/vkcom/src/context.hpp
@@ -12,8 +12,9 @@ namespace cv { namespace dnn { namespace vkcom {
 
 #ifdef HAVE_VULKAN
 
-struct Context
+class Context
 {
+public:
     Context();
     ~Context();
 };

--- a/modules/dnn/src/vkcom/src/op_base.cpp
+++ b/modules/dnn/src/vkcom/src/op_base.cpp
@@ -45,7 +45,9 @@ void OpBase::initVulkanThing(int buffer_num)
 
 void OpBase::createDescriptorSetLayout(int buffer_num)
 {
-    VkDescriptorSetLayoutBinding bindings[buffer_num] = {};
+    if (buffer_num <= 0)
+        return;
+    std::vector<VkDescriptorSetLayoutBinding> bindings(buffer_num);
     for (int i = 0; i < buffer_num; i++)
     {
         bindings[i].binding = i;
@@ -56,7 +58,7 @@ void OpBase::createDescriptorSetLayout(int buffer_num)
     VkDescriptorSetLayoutCreateInfo info = {};
     info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO;
     info.bindingCount = buffer_num;
-    info.pBindings = bindings;
+    info.pBindings = &bindings[0];
     VK_CHECK_RESULT(vkCreateDescriptorSetLayout(device_, &info, NULL, &descriptor_set_layout_));
 }
 

